Adds new tutorial to docs for deploying an EVM token exchange bridge on testnets (#4)
ChainSafe/chainbridge-docs: docs/live-evm-bridge.md

# Deploying a Live EVM->EVM Token Bridge

This tutorial walks through the process of deploying a token exchange bridge between two Ethereum test networks (Gorli and Rinkeby). It could similarly be applied to link any two Ethereum chains including mainnet.

At a high level setting up ChainBridge for token transfers requires the following:

- A token native to ChainA (e.g. TKN) and a destination token on ChainB which will represent the source token wrapped (e.g. wTKN)
- The ChainBridge bridge contract deployed to each chain
- Handler contracts on each chain to instruct the bridge what to do with the tokens
- One or more off-chain relayers running to pick up transactions from the source chain and relay them to the destination chain

## Overview of Cross-chain Token Transfers

> In reality it is impossible to actually transfer tokens between blockchains. That said, with a bit of contract magic it is easy to make something that is functionally equivalent.

Author: An extension of the idea that a blockchain can only alter its own state and nothing else and therefore tokens are inherently native to a single chain. A token transfer requires two state mutations (subtracting balance and adding to another account). If one account lives on another chain then this can't be done without some kind of locking, minting and relaying (contract magic :P).

Would it perhaps be better to say something like:

```suggestion
Tokens are inherently native to a single chain; however, with a bit of contract magic it is possible to make something equivalent to transferring value between two chains.
```

The bridge contract must be deployed on both the source and destination chains. Its primary task on the source chain is to broker token deposits (ensure they are locked) and emit the events that the relayers  listen for. On the destination chain it is responsible for managing a set of permissioned relayers, aggregating relayers votes on proposals passed from the source chain and executing the desired action (e.g. minting wrapped tokens) on the destination chain when the vote threshold is reached.

### Handlers

To allow extensibility the bridge contract is written to call functions in a handler contract when tokens are deposited on the source chain or when a proposal is approved on the destination chain. Each resource (e.g. token) can have a unique handler. You are free to write your own handler contract that implements any functionality you like. 

The ERC20 handler contract that ships with ChainBridge can be configured to either lock up the tokens or burn them (if the token allows) on deposit and either mint or release tokens when a proposal is executed. A single handler can handle many tokens but their contract addresses must be registered with the handler ahead of time.

> ### Relay

Author: Agreed this is more consistent

```suggestion
### Relayers
```

> The relay is an off-chain actor that listens for particular events on the source chain and when certain conditions are met will submit signed proposals to the destination chain. The public keys of the approved relays must be registered with the bridge contract on the destination chain.

Author: ```suggestion
A relayer is an off-chain actor that listens for particular events on the source chain and when certain conditions are met will submit signed proposals to the destination chain. The public keys of the approved relayers must be registered with the bridge contract on the destination chain.
```

Reviewer: Won't let me commit this, I think the changes overlapped. Lets get this in and merge this 👍

> A relay will also listen for new proposals on the destination chain to verify and cast votes for if valid. Once a proposal has sufficient votes a relayer can execute the proposal to trigger the handler.

Author: Ah cool. I thought they listened on both. Propose changing to: 
```suggestion
Once a proposal has sufficient votes a relayer can execute the proposal to trigger the handler.
```

You can use MetaMask to generate new accounts on the test networks and export your public and private keys.

Additional env vars are required to be set throughout the tutorial. If a command doesn't work be sure you have set the required variables.

### Steps

#### 1. Deploy contracts on Source (Goerli)

The following command will deploy the bridge contract and ERC20 handler contract on the source

```shell
cb-sol-cli --url $SRC_GATEWAY --privateKey $SRC_PK --gasPrice 10000000000 deploy \
    --bridge --erc20Handler \
    --relayers $SRC_ADDR \
    --relayerThreshold 1\
    --chainId 0
```

Take note of the output of the above command and assign the following variables

```shell
SRC_BRIDGE="<resulting bridge contract address>"
SRC_HANDLER="<resulting erc20 handler contract address>"
```

#### 2. Configure contracts on Source

The following registers the WEENUS token as a resource with on bridge contract and configures which handler to use

```shell
cb-sol-cli --url $SRC_GATEWAY --privateKey $SRC_PK --gasPrice 10000000000 bridge register-resource \
    --bridge $SRC_BRIDGE \
    --handler $SRC_HANDLER \
    --resourceId $RESOURCE_ID \
    --targetContract $SRC_TOKEN
```

> *Note: Sometimes the transaction confirmation will not be detected by the CLI. You can check the txhash in etherscan to see if it has been confirmed.*

Reviewer: Interesting... it has been pretty dependable in my experience. If you experience this again please open an issue on chainbridge-deploy.

The bridge contract must be deployed on both the source and destination chains. Its primary task on the source chain is to broker token deposits (ensure they are locked) and emit the events that the relayers listen for. On the destination chain it is responsible for managing a set of permissioned relayers, aggregating relayer votes on proposals passed from the source chain and executing the desired action (e.g. minting tokens) on the destination chain when the vote threshold is reached.

### Handlers

To allow extensibility the bridge contract is written to call functions in a handler contract when tokens are deposited on the source chain or when a proposal is approved on the destination chain. Each resource (e.g. token) can have a unique handler. You are free to write your own handler contract that implements any functionality you like. 

The ERC20 handler contract that ships with ChainBridge can be configured to either lock up the tokens or burn them (if the token allows) on deposit and either mint or release tokens when a proposal is executed. A single handler can handle many tokens but their contract addresses must be registered with the handler ahead of time.

### Relayers

> The relay is an off-chain actor that listens for particular events on the source chain and when certain conditions are met will submit signed proposals to the destination chain. The addresses of the approved relays must be registered with the bridge contract on the destination chain.

Author: ```suggestion
The relayer is an off-chain actor that listens for particular events on the source chain and when certain conditions are met will submit signed proposals to the destination chain. The addresses of the approved relays must be registered with the bridge contract on the destination chain.
```
